lib/rules/padding-line-between-statements.js

```diff
@@ -0,0 +1,585 @@
+// Helpers
+//------------------------------------------------------------------------------
+
+const VALID_PARENT = /^(?:Program|(?:Switch|Block)Statement|SwitchCase)$/;
+const LT = `(?:${astUtils.LINEBREAK_MATCHER.source})`;
+const PADDING_LINE_SEQUENCE = new RegExp(
+    String.raw`^(\s*?${LT})\s*${LT}(\s*;?)$`
+);
+const CJS_EXPORT = /^(?:module\s*\.\s*)?exports(?:\s*\.|\s*\[|$)/;
+const CJS_IMPORT = /^require\(/;
+
+/**
+ * Creates tester which check if a node starts with specific keyword.
+ *
+ * @param {string} keyword The keyword to test.
+ * @returns {Object} the created tester.
+ * @private
+ */
+function newKeywordTester(keyword) {
+    return {
+        test: (node, sourceCode) =>
+            sourceCode.getFirstToken(node).value === keyword
+    };
+}
+
+/**
+ * Creates tester which check if a node is specific type.
+ *
+ * @param {string} type The node type to test.
+ * @returns {Object} the created tester.
+ * @private
+ */
+function newNodeTypeTester(type) {
+    return {
+        test: node =>
+            node.type === type
+    };
+}
+
+/**
+ * Checks the given node is an expression statement of IIFE.
+ *
+ * @param {ASTNode} node The node to check.
+ * @returns {boolean} `true` if the node is an expression statement of IIFE.
+ * @private
+ */
+function isIIFEStatement(node) {
+    return (
+        node.type === "ExpressionStatement" &&
+        node.expression.type === "CallExpression" &&
+        astUtils.isFunction(node.expression.callee)
+    );
+}
+
+/**
+ * Checks whether the given node is a block-like statement.
+ * This checks the last token of the node is the closing brace of a block.
+ *
+ * @param {SourceCode} sourceCode The source code to get tokens.
+ * @param {ASTNode} node The node to check.
+ * @returns {boolean} `true` if the node is a block-like statement.
+ * @private
+ */
+function isBlockLikeStatement(sourceCode, node) {
+
+    // do-while with a block is a block-like statement.
+    if (node.type === "DoWhileStatement" && node.body.type === "BlockStatement") {
+        return true;
+    }
+
+    // IIFE is a block-like statement specially from
+    // JSCS#disallowPaddingNewLinesAfterBlocks.
+    if (isIIFEStatement(node)) {
+        return true;
+    }
+
+    // Checks the last token is a closing brace of blocks.
+    const lastToken = sourceCode.getLastToken(node, astUtils.isNotSemicolonToken);
+    const belongingNode = astUtils.isClosingBraceToken(lastToken)
+        ? sourceCode.getNodeByRangeIndex(lastToken.range[0])
+        : null;
+
+    return Boolean(belongingNode) && (
+        belongingNode.type === "BlockStatement" ||
+        belongingNode.type === "SwitchStatement"
+    );
+}
+
+/**
+ * Check whether the given node is a directive or not.
+ * @param {ASTNode} node The node to check.
+ * @param {SourceCode} sourceCode The source code object to get tokens.
+ * @returns {boolean} `true` if the node is a directive.
+ */
+function isDirective(node, sourceCode) {
```

Comment:
I thought it's not efficient if I make the array of directive prologue then do linear search.